src/uct/cuda/cuda_copy/cuda_copy_iface.c

```diff
@@ -86,27 +86,104 @@ static ucs_status_t uct_cuda_copy_iface_query(uct_iface_h iface,
     return UCS_OK;
 }
 
+static ucs_status_t uct_cuda_copy_iface_flush(uct_iface_h tl_iface, unsigned flags,
+                                              uct_completion_t *comp)
+{
+    uct_cuda_copy_iface_t *iface = ucs_derived_of(tl_iface, uct_cuda_copy_iface_t);
+
+    if (comp != NULL) {
+        return UCS_ERR_UNSUPPORTED;
+    }
+
+    if (ucs_queue_is_empty(&iface->outstanding_cuda_event_q)) {
+        UCT_TL_IFACE_STAT_FLUSH(ucs_derived_of(tl_iface, uct_base_iface_t));
+        return UCS_OK;
+    }
+
+    UCT_TL_IFACE_STAT_FLUSH_WAIT(ucs_derived_of(tl_iface, uct_base_iface_t));
+    return UCS_INPROGRESS;
+}
+
+static unsigned uct_cuda_copy_iface_progress(uct_iface_h tl_iface)
+{
+    uct_cuda_copy_iface_t *iface = ucs_derived_of(tl_iface, uct_cuda_copy_iface_t);
+    unsigned count = 0;
+    cudaError_t result = cudaSuccess;
+    uct_cuda_copy_event_desc_t *cuda_event;
+    ucs_queue_iter_t iter;
+
+    ucs_queue_for_each_safe(cuda_event, iter, &iface->outstanding_cuda_event_q, queue) {
```

Comment:
I don't like going over all outstanding events here. Is it  possible to do it more efficiently ?

Reply:
@alex-mikheev i couldn't come up any better now. for correctness,  I would like go with this now. will revisit this again when doing protocol performance evaluations.

Reply:
Here http://on-demand.gputechconf.com/gtc/2014/presentations/S4158-cuda-streams-best-practices-common-pitfalls.pdf it says that the operations in the same stream are ordered and can not overlap. If this is true then checking completion status of the first element of the outstanding queue is enough. Later it is possible to have separate outstanding queues for put and get.

Reply:
@alex-mikheev Yes, same stream events are ordered. I split into separate queues for h2d and d2h